HAR-1651: ajastetut tehtävät tietokantataulu

solita-laurin opened the pull request:
Tietokantataulu ajastetuille tehtäville, johon merkitään viimeisin yritys ja viimeisin onnitunut tehtävä

solita-laurin committed:
Tietokantataulu

solita-laurin committed:
Merge branch 'develop' of https://github.com/finnishtransportagency/harja into HAR-1651

solita-laurin committed:
Tietokantataulu ajastetuille tehtäville, johon merkitään viimeisin yritys ja viimeisin onnitunut tehtävä

solita-laurin committed:
Merge branch 'develop' of https://github.com/finnishtransportagency/harja into HAR-1651

rckopteri commented on src/clj/harja/kyselyt/ajastetut_tehtavat.sql:
```diff
@@ -0,0 +1,8 @@
+-- name: paivita-ajastetun-tehtavan-onnistuminen!
```
SQL formatoinnissa Harjassa on pyritty käyttämään "river" tyyppistä formatointia: https://www.sqlstyle.guide/

rckopteri commented on src/clj/harja/kyselyt/ajastetut_tehtavat.sql:
```diff
@@ -0,0 +1,8 @@
+-- name: paivita-ajastetun-tehtavan-onnistuminen!
+UPDATE ajastetut_tehtavat
+SET suoritusyritys_aika = NOW(), onnistunut = :onnistunut
+WHERE nimi = 'siirra_toteumat_analytiikalle';
```
Funktion nimi "paivita-ajastetun-tehtavan-onnistuminen!" antaa ymmärtää, että funktiolla voi päivittää minkä tahansa tehtävän onnistuneeksi. Nyt se on kuitenkin kovakoodattu "siirra_toteumat_analytiikalle" tyyppiseksi. Generaalimpi nimi on hyvä, mutta anna tyyppi myös parametrina. Sama homma tuossa alemmassa funkkarissa.

rckopteri commented on src/clj/harja/kyselyt/ajastetut_tehtavat.clj:
```diff
@@ -0,0 +1,7 @@
+(ns harja.kyselyt.ajastetut-tehtavat
```
Harjan Readmessä  on tiimin täätöksenä 9/2021 määritelty, että kyselytiedostojen nimessä on aina _kyselyt.clj pääte

rckopteri commented on src/clj/harja/palvelin/integraatiot/api/analytiikka.clj:
```diff
@@ -283,7 +283,7 @@
                            (konversio/alaviiva->rakenne toteuma))
                       toteumat)})
-        _ (toteuma-kyselyt/paivita-palautettu-analytiikalle-aikaleima! db {:alkuaika alkuaika
+        #_#__ (toteuma-kyselyt/paivita-palautettu-analytiikalle-aikaleima! db {:alkuaika alkuaika
```
Jos tätä ei tarvita, niin poista koodi vain kokonaan.

solita-laurin replied:
Tää kommentointiin pois, koska Maarit löysi onglema tuotannonpuolella jossa on readreplicakanta, johon vaatii lisäkorjauksen

solita-laurin replied:
Tukitiiminpalaverissa katottiin tätä, päätettiin että olisi isompi korjaus tähän, joten tehtiin erillinen tiketti myöhemmäksi.  Poistan tän rivin nytten.

rckopteri commented on tietokanta/src/main/resources/db/migration/V1_1216__.sql:
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE ajastetut_tehtavat
+(
+    nimi                    varchar(256) not null,
```
Nimen sijasta minusta voisi käyttää "tyyppi" nimistä kenttää ja tehdä siitä ENUM- tyyppinen.

rckopteri commented on tietokanta/src/main/resources/db/migration/V1_1216__.sql:
```diff
@@ -0,0 +1,9 @@
+    viimeisin_onnistunut    timestamp
+);
+
+INSERT INTO ajastetut_tehtavat (nimi) VALUES ('siirra_toteumat_analytiikalle');
```
Tätä ei varmastikaan tarvita, kun tässä ei määritellä vielä, että jokin ajastettu tehtävä olisi merkattu ajastettuna.

solita-laurin committed:
Tietokantataulukyselyn generalisointi ja nimen muutokset

solita-laurin committed:
Merge branch 'develop' of https://github.com/finnishtransportagency/harja into HAR-1651

solita-laurin committed:
Tietokantafunktion muutos testien korjaukselle ja paremmalle ajankäsittelylle

solita-laurin committed:
Merge branch 'develop' into HAR-1651

solita-laurin committed:
Merge branch 'develop' of https://github.com/finnishtransportagency/harja into HAR-1651

solita-laurin committed:
merge-fix

solita-laurin committed:
Merge branch 'develop' of https://github.com/finnishtransportagency/harja into HAR-1651

solita-laurin committed:
Merge branch 'HAR-1651' of https://github.com/finnishtransportagency/harja into HAR-1651

solita-laurin committed:
merge-conflic fix

maaritlaine commented on tietokanta/src/main/resources/db/migration/V1_1217__.sql:
```diff
@@ -0,0 +1,12 @@
+CREATE TYPE ajastetuntehtavan_tyyppi AS ENUM ('siirra_toteumat_analytiikalle');
```
Muuta tyypin nimi ajastetun_tehtavan_tyyppi, niin nimi on nopeempi hahmottaa. Ajastettu tehtävä ei oo yhdyssana.

maaritlaine commented on src/clj/harja/palvelin/ajastetut_tehtavat/analytiikan_toteumat.clj:
```diff
@@ -14,8 +15,12 @@
+        nyt (or annettu-nyt nil)]
+    (try
+        (toteuma-kyselyt/siirra-toteumat-analytiikalle db {:nyt (konversio/sql-timestamp nyt)})
+        (ajastetut-tehtavat-kyselyt/paivita-ajastetun-tehtavan-onnistuminen! db true "siirra_toteumat_analytiikalle")
```
Täällä voi vähän yksinkertaistaa asioita.

maaritlaine commented on tietokanta/src/main/resources/db/migration/R__Analytiikka.sql:
```diff
@@ -42,8 +42,12 @@ BEGIN
                     JOIN urakka u ON t.urakka = u.id
-          WHERE (t.luotu BETWEEN ajankohta - '1 day'::INTERVAL AND ajankohta)
-             OR (tm.luotu BETWEEN ajankohta - '1 day'::INTERVAL AND ajankohta)
+         WHERE (t.luotu BETWEEN (SELECT COALESCE(ajankohta, viimeisin_onnistunut, NOW() - '1 day'::INTERVAL)
```
Näiden ehtojen kohdalla kannattaa miettiä ja kokeilla, mitä tietokannasta palautuu. 

Me halutaan, että jos kutsuva taho ei ole antanut **tiettyä ajankohtaa** parametreissa, kysely palauttaa ne toteumat, jotka on luotu tai joita on muokattu sen jälkeen, kun ajastus on **edellisen kerran onnistuneesti** pyörähtänyt. Jos viimeisintä onnistunutta ajankohtaa ei ole, onkin hyvä idea käyttää vanhaa tapaa eli **nyt - 1 päivä** -aikaleimaa.

maaritlaine reviewed (changes requested):
Kommentit tuolla koodin seassa

rckopteri committed:
Merge branch 'develop' into HAR-1651

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_1217__.sql

rckopteri committed:
Muokkaa toteumien siirto analytiikka_toteumat tauluun toimimaan alku- ja loppuajan perusteella.

Muokkaa yksikkötesti vastaamaan tätä. Käytä myös hyväksi edellistä logitettua siirtoajankohtaa ajastetut_tehtavat taulusta.

rckopteri committed:
Lisää kolme uudelleen yritystä siirtoon ja varmista, että siirron ikkuna on kolme tuntia suurempi, kuin mitä käyttäjä olettaa.

Eli varmistetaan, että kaikki toteumat todella siirtyy

rckopteri committed:
Merge branch 'develop' into HAR-1651

rckopteri committed:
Merge branch 'develop' into HAR-1651

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_1244__.sql

rckopteri committed:
Korjaa virheellinen haku

rckopteri committed:
Päivitä loputkin testit pelaamaan uudistetun analytikka_toteumat siirron kanssa

maaritlaine commented on test/clj/harja/palvelin/ajastetut_tehtavat/analytiikan_toteumat_siirto_test.clj:
```diff
@@ -35,25 +25,103 @@
+        "Siirron jälkeen pitäisi olla täsmälleen yksi ajastetut_tehtavat -rivi")
+
+    ;; Tarkista, että lokirivissä on oikea tyyppi
+    (is (= "siirra_toteumat_analytiikalle" (:tyyppi logi))
```
Tyypillä rajataan jo hakuehdoissa. Tämä testi ei voi koskaan feilata

maaritlaine commented on test/clj/harja/palvelin/ajastetut_tehtavat/analytiikan_toteumat_siirto_test.clj:
```diff
@@ -35,25 +25,103 @@
+        loppupaiva (pvm/luo-pvm 2015 11 31)
         ;; HAetaan vain vuoden 2015 toteumat ja siivotaan varalta kaikki 2015 vuoden mahdollisesti siirretyt toteumat pois
         _ (u "DELETE FROM analytiikka_toteumat WHERE toteuma_alkanut > '2014-12-31' AND toteuma_alkanut < '2016-01-01'")
+        _ (u "DELETE FROM ajastetut_tehtavat WHERE suoritusyritys_aika > '2014-12-31' AND suoritusyritys_aika < '2016-01-01'")
```
Tuossa suoritusyritys_aika-kentän käytössä on sekaantumisen vaara, jos siihen tallennetaan viimeksi siirrettyyn ajanjaksoon liittyvä aikaleima. Tartteeko toiminnallisuuden oikeesti olla testissä ja lokaalissa erilainen kuin tuotannossa?

Jos halutaan, että ajastuksen voi määrittää siirtämään jokin tietty, väilin jäänyt tai virheellisesti siirtynyt ajanjakso (ei välttämättä mikään huono ajatus), kannattaisi ehkä lisätä tietokantaan suoritusyritys_aika-kentän lisäksi sarakkeet haettavan ajanjakson alku- ja loppuajoille. Jos ne on asetettu, hyödynnetään niitä. Jos ne puuttuu, siirretään suoritusyritys_aika - nyt väli.

rckopteri committed:
Refaktorointia katselmointikommenttien perusteella.

Jakso, jolle automaattinen tehtävä ajetaan tallennetaan nyt kantaan.

rckopteri committed:
Merge branch 'develop' into HAR-1651

rckopteri committed:
Poista muutama turha println testeistä

rckopteri committed:
Korjaa testi

solita-staskila commented on src/clj/harja/palvelin/ajastetut_tehtavat/analytiikan_toteumat.clj:
```diff
@@ -3,19 +3,80 @@
+        viimeisin-ajokerta (when viimeisin-ajokerta (pvm/ajan-muokkaus viimeisin-ajokerta false 3 :tunti))
+
+        annettu-alkuaika (first args)                       ;; Annetaan testeissä
+        annettu-loppuaika (second args)                     ;; Annetaan testeissä
```
Ei tehdä tehdä prodi koodiin lisää logiikkaa joka on pelkästään dedikoitu testeille. Jos argumentit on pelkästään testejä varten, siivotaan ne tästä pois. Funktion tulisi tehdä vain se, mitä sen on tarkoitus tehdä, ja testit observoi, eläen aivan omaa elämää. 

Jos tässä on tarkoitus antaa viimeisin ajokerta, eikö yksikkötesti voi vaan inserttaa/päivittää suoraan ajastetut_tehtavat tauluun tuon timestampin, sitten ajaa tämä funktio ilman tällaisia extra argumentteja? Tätä pitää miettiä vähän uudelleen

rckopteri replied:
Miellyttävä ajatus. Mäpä muutaman tämän 👍

solita-staskila commented on src/clj/harja/palvelin/ajastetut_tehtavat/analytiikan_toteumat.clj:
```diff
@@ -3,19 +3,80 @@
+                   viimeisin-ajokerta                       ;; Saatiin tietokanansta
+                   (pvm/ajan-muokkaus pvm/nyt true 1 :paiva) ;; Ensimmäisellä kerralla siirretään viimeisen päivän toteumat
+                   )
+        alkuaika-sql (if (= "java.sql.Timestamp" (type alkuaika)) alkuaika (konversio/sql-timestamp alkuaika))
```
Ei tarvi olla keissiä jolloin alkuaika on jotain muuta tyyppiä, tämä myös luultavasti noiden testiargumenttien takia?

rckopteri replied:
Joo. Tämä yksinkertaistui samalla

solita-henrikko commented on tietokanta/src/main/resources/db/migration/V1_1245__.sql:
Versionumero pitää päivittää, muuten näyttää hyvältä

solita-staskila commented on src/clj/harja/palvelin/ajastetut_tehtavat/analytiikan_toteumat.clj:
```diff
@@ -3,19 +3,80 @@
+        ;; Kovakoodattu maksimiyritysmäärä. Ei tarvetta vielä asetuksiin.
+        max-yritykset 3]
+    (loop [yritys 1
+           viimeisin-virhe nil]
```
Ollaankos me selvitetty syy, mistä errori tulee? 

Koska jos tietokanta kysely failaa, miksi se onnistuisi toisella yrityksellä? 

Eli onko meillä tapahtunut jotain connection pool faileja, deadlockkeja tms? 

Jos ei ole tietoa, ottaisin tämän koko retry logiikan pois kokonaan, ja logitetaan ensin vaan virhe. 

Sitten kun tiedetään virhe, voidaan päätellä onko tällainen uudelleenyritys tarpeellinen 

Jos tässä tapahtuu jokin nullpointerexception, silloin koodissa on vika, eikä tämä retry aja mitään käytännön tarkoitusta

rckopteri replied:
Mä en tiedä ainoastakaan failista. Olikohan @maaritlaine :lla muuta tietoa?

solita-staskila replied:
Käykö, jos otetaan tuo retry logiikka sitten pois?

Se on aika teippi-ratkaisu, eikä addressaa oikeaa ongelmaa. (sori, jos kuulostan kritisoivalta :smile: tarkoitus vaan antaa näkemystä) 

Luetaan sitten logilta, mitä virhettä sieltä tulee, ja tehdään sen perusteella toimenpiteitä

rckopteri committed:
Merge branch 'develop' into HAR-1651

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_1245__.sql

rckopteri committed:
Poista aikaparametrit siirtofunktiosta ja muokkaa testejä niin, että voidaan päätellä samalla tavalla

että toteumien siirto sujuu, olipa siirtoa tehty aiemmin tai ei.

rckopteri committed:
Merge branch 'develop' into HAR-1651

# Conflicts:
#	test/clj/harja/palvelin/palvelut/suunnittelu/tarjous_test.clj
#	tietokanta/src/main/resources/db/migration/V1_1249__.sql

rckopteri committed:
Poista uudelleen yritykset toteumien siirrosta

rckopteri committed:
Poista turha konversio
